File: dart/zed_sync/lib/src/lifecycle.dart
/// Total, deterministic application/session lifecycle for Flutter apps.
///
/// Platform effects run only after an [applied] transition. Every asynchronous
/// completion carries the operation token allocated by its request; an older
/// completion is [stale] and cannot resurrect a stopped session.
library;

enum AppPhase { stopped, starting, online, offline, stopping, failed }

enum LifecycleOperation { none, start, stop, reconcile }

enum LifecycleFailure { start, runtime, stop }

enum TransitionOutcome { applied, stuttered, stale, rejected }

enum _TokenRelation { current, stale, invalid }

sealed class LifecycleEvent {
  const LifecycleEvent();
}

final class StartRequested extends LifecycleEvent {
  const StartRequested();
}

final class StartSucceeded extends LifecycleEvent {
  final int token;
  const StartSucceeded(this.token);
}

final class StartFailed extends LifecycleEvent {
  final int token;
  const StartFailed(this.token);
}

final class ConnectivityChanged extends LifecycleEvent {
  final int token;
  final bool online;
  const ConnectivityChanged(this.token, this.online);
}

final class RuntimeFailed extends LifecycleEvent {
  final int token;
  const RuntimeFailed(this.token);
}

final class StopRequested extends LifecycleEvent {
  const StopRequested();
}

final class StopSucceeded extends LifecycleEvent {
  final int token;
  const StopSucceeded(this.token);
}

final class StopFailed extends LifecycleEvent {
  final int token;
  const StopFailed(this.token);
}

final class ReconcileRequested extends LifecycleEvent {
  const ReconcileRequested();
}

class AppLifecycleSnapshot {
  final AppPhase phase;
  final LifecycleOperation operation;
  final int generation;
  final int? activeToken;
  final bool desiredRunning;
  final bool online;
  final LifecycleFailure? failure;

  const AppLifecycleSnapshot({
    this.phase = AppPhase.stopped,
    this.operation = LifecycleOperation.none,
    this.generation = 0,
    this.activeToken,
    this.desiredRunning = false,
    this.online = false,
    this.failure,
  });

  AppCapabilities get capabilities => AppCapabilities._(this);

  Map<String, dynamic> toJson() => {
        'phase': phase.name,
        'operation': operation.name,
        'generation': generation,
        'active_token': activeToken,
        'desired_running': desiredRunning,
        'online': online,
        'failure': failure?.name,
      };

  /// Check the phase/operation/token/failure coherence proved by Quint.
  void validate() {
    final tokenValid =
        activeToken == null || (activeToken! > 0 && activeToken! <= generation);
    if (generation < 0 ||
        generation > AppLifecycleMachine.maxSafeGeneration ||
        !tokenValid) {
      throw StateError(
          'zed-sync lifecycle: invalid generation or active token');
    }

    final coherent = switch (phase) {
      AppPhase.stopped => operation == LifecycleOperation.none &&
          activeToken == null &&
          !desiredRunning &&
          !online &&
          failure == null,
      AppPhase.starting => operation == LifecycleOperation.start &&
          activeToken == generation &&
          desiredRunning &&
          failure == null,
      AppPhase.online => operation == LifecycleOperation.none &&
          activeToken == generation &&
          desiredRunning &&
          online &&
          failure == null,
      AppPhase.offline => operation == LifecycleOperation.none &&
          activeToken == generation &&
          desiredRunning &&
          !online &&
          failure == null,
      AppPhase.stopping => (operation == LifecycleOperation.stop ||
              operation == LifecycleOperation.reconcile) &&
          activeToken == generation &&
          !desiredRunning &&
          !online &&
          failure == null,
      AppPhase.failed => operation == LifecycleOperation.none &&
          activeToken == null &&
          !desiredRunning &&
          !online &&
          failure != null,
    };
    if (!coherent) {
      throw StateError(
          'zed-sync lifecycle: phase, operation, token, intent, connectivity, and failure disagree');
    }
  }
}

class AppCapabilities {
  final bool canStart;
  final bool canStop;
  final bool canWrite;
  final bool canReceiveChanges;
  final bool canFlush;
  final bool canReconcile;
  final bool busy;
  final bool running;

  AppCapabilities._(AppLifecycleSnapshot snapshot)
      : this._fromPhase(_validatedPhase(snapshot));

  AppCapabilities._fromPhase(AppPhase? phase)
      : canStart = phase == AppPhase.stopped,
        canStop = const {AppPhase.starting, AppPhase.online, AppPhase.offline}
            .contains(phase),
        canWrite = const {AppPhase.online, AppPhase.offline}.contains(phase),
        canReceiveChanges = const {
          AppPhase.starting,
          AppPhase.online,
          AppPhase.offline
        }.contains(phase),
        canFlush = phase == AppPhase.online,
        canReconcile = phase == AppPhase.failed,
        busy = const {AppPhase.starting, AppPhase.stopping}.contains(phase),
        running = const {AppPhase.online, AppPhase.offline}.contains(phase);

  static AppPhase? _validatedPhase(AppLifecycleSnapshot snapshot) {
    try {
      snapshot.validate();
      return snapshot.phase;
    } catch (_) {
      return null;
    }
  }
}

typedef LifecycleListener = void Function(
    AppLifecycleSnapshot snapshot, LifecycleEvent event);

class AppLifecycleMachine {
  static const maxSafeGeneration = 9007199254740991;

  AppLifecycleSnapshot _snapshot = const AppLifecycleSnapshot();
  final List<LifecycleListener> _listeners = [];

  AppLifecycleSnapshot get snapshot => _snapshot;
  AppCapabilities get capabilities => _snapshot.capabilities;

  /// Subscribe to applied transitions. The returned callback unsubscribes
  /// idempotently; observer exceptions never control state.
  void Function() subscribe(LifecycleListener listener) {
    _listeners.add(listener);
    var subscribed = true;
    return () {
      if (!subscribed) return;
      subscribed = false;
      _listeners.remove(listener);
    };
  }

  TransitionOutcome dispatch(LifecycleEvent event) {
    final before = _snapshot;
    AppLifecycleSnapshot? next;
    var outcome = TransitionOutcome.rejected;

    if (event is StartRequested) {
      if (before.phase == AppPhase.stopped &&
          before.generation < maxSafeGeneration) {
        final token = before.generation + 1;
        next = AppLifecycleSnapshot(
          phase: AppPhase.starting,
          operation: LifecycleOperation.start,
          generation: token,
          activeToken: token,
          desiredRunning: true,
        );
        outcome = TransitionOutcome.applied;
      }
    } else if (event is StartSucceeded) {
      switch (_tokenRelation(event.token)) {
        case _TokenRelation.stale:
          return TransitionOutcome.stale;
        case _TokenRelation.invalid:
          return TransitionOutcome.rejected;
        case _TokenRelation.current:
          break;
      }
      if (before.phase == AppPhase.starting) {
        next = _copy(
          before,
          phase: before.online ? AppPhase.online : AppPhase.offline,
          operation: LifecycleOperation.none,
        );
        outcome = TransitionOutcome.applied;
      }
    } else if (event is StartFailed) {
      switch (_tokenRelation(event.token)) {
        case _TokenRelation.stale:
          return TransitionOutcome.stale;
        case _TokenRelation.invalid:
          return TransitionOutcome.rejected;
        case _TokenRelation.current:
          break;
      }
      if (before.phase == AppPhase.starting) {
        next = _failed(LifecycleFailure.start);
        outcome = TransitionOutcome.applied;
      }
    } else if (event is ConnectivityChanged) {
      switch (_tokenRelation(event.token)) {
        case _TokenRelation.stale:
          return TransitionOutcome.stale;
        case _TokenRelation.invalid:
          return TransitionOutcome.rejected;
        case _TokenRelation.current:
          break;
      }
      if (const {AppPhase.starting, AppPhase.online, AppPhase.offline}
          .contains(before.phase)) {
        final phase = before.phase == AppPhase.starting
            ? AppPhase.starting
            : event.online
                ? AppPhase.online
                : AppPhase.offline;
        if (before.online == event.online && before.phase == phase) {
          outcome = TransitionOutcome.stuttered;
        } else {
          next = _copy(before, phase: phase, online: event.online);
          outcome = TransitionOutcome.applied;
        }
      }
    } else if (event is RuntimeFailed) {
      switch (_tokenRelation(event.token)) {
        case _TokenRelation.stale:
          return TransitionOutcome.stale;
        case _TokenRelation.invalid:
          return TransitionOutcome.rejected;
        case _TokenRelation.current:
          break;
      }
      if (const {AppPhase.starting, AppPhase.online, AppPhase.offline}
          .contains(before.phase)) {
        next = _failed(LifecycleFailure.runtime);
        outcome = TransitionOutcome.applied;
      }
    } else if (event is StopRequested) {
      if (const {AppPhase.stopped, AppPhase.stopping}.contains(before.phase)) {
        outcome = TransitionOutcome.stuttered;
      } else if (const {AppPhase.starting, AppPhase.online, AppPhase.offline}
          .contains(before.phase)) {
        next = _beginStop(LifecycleOperation.stop);
        outcome = next == null
            ? TransitionOutcome.rejected
            : TransitionOutcome.applied;
      }
    } else if (event is StopSucceeded) {
      switch (_tokenRelation(event.token)) {
        case _TokenRelation.stale:
          return TransitionOutcome.stale;
        case _TokenRelation.invalid:
          return TransitionOutcome.rejected;
        case _TokenRelation.current:
          break;
      }
      if (before.phase == AppPhase.stopping) {
        next = AppLifecycleSnapshot(generation: before.generation);
        outcome = TransitionOutcome.applied;
      }
    } else if (event is StopFailed) {
      switch (_tokenRelation(event.token)) {
        case _TokenRelation.stale:
          return TransitionOutcome.stale;
        case _TokenRelation.invalid:
          return TransitionOutcome.rejected;
        case _TokenRelation.current:
          break;
      }
      if (before.phase == AppPhase.stopping) {
        next = _failed(LifecycleFailure.stop);
        outcome = TransitionOutcome.applied;
      }
    } else if (event is ReconcileRequested) {
      if (before.phase == AppPhase.failed) {
        next = _beginStop(LifecycleOperation.reconcile);
        outcome = next == null
            ? TransitionOutcome.rejected
            : TransitionOutcome.applied;
      }
    }

    if (outcome == TransitionOutcome.applied) _commit(next!, event);
    return outcome;
  }

  _TokenRelation _tokenRelation(int token) => switch (token) {
        <= 0 => _TokenRelation.invalid,
        final candidate when _snapshot.activeToken == candidate =>
          _TokenRelation.current,
        final candidate when candidate <= _snapshot.generation =>
          _TokenRelation.stale,
        _ => _TokenRelation.invalid,
      };

  AppLifecycleSnapshot? _beginStop(LifecycleOperation operation) {
    if (_snapshot.generation >= maxSafeGeneration) return null;
    final token = _snapshot.generation + 1;
    return AppLifecycleSnapshot(
      phase: AppPhase.stopping,
      operation: operation,
      generation: token,
      activeToken: token,
    );
  }

  AppLifecycleSnapshot _failed(LifecycleFailure failure) =>
      AppLifecycleSnapshot(
        phase: AppPhase.failed,
        generation: _snapshot.generation,
        failure: failure,
      );

  AppLifecycleSnapshot _copy(
    AppLifecycleSnapshot source, {
    AppPhase? phase,
    LifecycleOperation? operation,
    bool? online,
  }) =>
      AppLifecycleSnapshot(
        phase: phase ?? source.phase,
        operation: operation ?? source.operation,
        generation: source.generation,
        activeToken: source.activeToken,
        desiredRunning: source.desiredRunning,
        online: online ?? source.online,
        failure: source.failure,
      );

  void _commit(AppLifecycleSnapshot next, LifecycleEvent event) {
    next.validate();
    _snapshot = next;
    for (final listener in List<LifecycleListener>.of(_listeners)) {
      try {
        listener(next, event);
      } catch (_) {
        // Observation must never control a lifecycle transition.
      }
    }
  }
}

File: src/lifecycle.rs
//! Total, deterministic application/session lifecycle shared by every app runtime.
//!
//! The reducer owns operational intent. Callers execute platform effects only
//! after an `Applied` transition and attach every asynchronous completion to
//! the returned `active_token`. A completion from an older generation is
//! classified as `Stale` and cannot resurrect a stopped or replaced session.

use serde::{Deserialize, Serialize};

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum AppPhase {
    Stopped,
    Starting,
    Online,
    Offline,
    Stopping,
    Failed,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LifecycleOperation {
    None,
    Start,
    Stop,
    Reconcile,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum LifecycleFailure {
    Start,
    Runtime,
    Stop,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub enum TransitionOutcome {
    Applied,
    Stuttered,
    Stale,
    Rejected,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum LifecycleEvent {
    StartRequested,
    StartSucceeded { token: u64 },
    StartFailed { token: u64 },
    ConnectivityChanged { token: u64, online: bool },
    RuntimeFailed { token: u64 },
    StopRequested,
    StopSucceeded { token: u64 },
    StopFailed { token: u64 },
    ReconcileRequested,
}

/// Authority relation between a callback token and the lifecycle generation.
///
/// Only a previously allocated, now-revoked token is stale. Zero and tokens
/// from an unallocated future generation are invalid and must be rejected.
#[derive(Clone, Copy, Debug, PartialEq, Eq)]
enum TokenRelation {
    Current,
    Stale,
    Invalid,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
pub struct AppLifecycleSnapshot {
    pub phase: AppPhase,
    pub operation: LifecycleOperation,
    pub generation: u64,
    pub active_token: Option<u64>,
    pub desired_running: bool,
    pub online: bool,
    pub failure: Option<LifecycleFailure>,
}

impl Default for AppLifecycleSnapshot {
    fn default() -> Self {
        Self {
            phase: AppPhase::Stopped,
            operation: LifecycleOperation::None,
            generation: 0,
            active_token: None,
            desired_running: false,
            online: false,
            failure: None,
        }
    }
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub struct AppCapabilities {
    pub can_start: bool,
    pub can_stop: bool,
    pub can_write: bool,
    pub can_receive_changes: bool,
    pub can_flush: bool,
    pub can_reconcile: bool,
    pub busy: bool,
    pub running: bool,
}

impl AppCapabilities {
    pub const CLOSED: Self = Self {
        can_start: false,
        can_stop: false,
        can_write: false,
        can_receive_changes: false,
        can_flush: false,
        can_reconcile: false,
        busy: false,
        running: false,
    };
}

impl AppLifecycleSnapshot {
    pub fn capabilities(&self) -> AppCapabilities {
        if self.validate().is_err() {
            return AppCapabilities::CLOSED;
        }

        let mut capabilities = AppCapabilities::CLOSED;
        match self.phase {
            AppPhase::Stopped => capabilities.can_start = true,
            AppPhase::Starting => {
                capabilities.can_stop = true;
                capabilities.can_receive_changes = true;
                capabilities.busy = true;
            }
            AppPhase::Online => {
                capabilities.can_stop = true;
                capabilities.can_write = true;
                capabilities.can_receive_changes = true;
                capabilities.can_flush = true;
                capabilities.running = true;
            }
            AppPhase::Offline => {
                capabilities.can_stop = true;
                capabilities.can_write = true;
                capabilities.can_receive_changes = true;
                capabilities.running = true;
            }
            AppPhase::Stopping => capabilities.busy = true,
            AppPhase::Failed => capabilities.can_reconcile = true,
        }
        capabilities
    }

    /// Check the phase/operation/token/failure coherence proved by the finite
    /// Quint model. This remains public so embedding apps can fail closed when
    /// decoding a persisted or foreign snapshot.
    pub fn validate(&self) -> Result<(), &'static str> {
        if self.generation > AppLifecycleMachine::MAX_SAFE_GENERATION {
            return Err("generation exceeds the shared exact-integer range");
        }
        if let Some(token) = self.active_token {
            if token == 0 || token > self.generation {
                return Err("active token must identify an allocated generation");
            }
        }
        let coherent = match self.phase {
            AppPhase::Stopped => {
                self.operation == LifecycleOperation::None
                    && self.active_token.is_none()
                    && !self.desired_running
                    && !self.online
                    && self.failure.is_none()
            }
            AppPhase::Starting => {
                self.operation == LifecycleOperation::Start
                    && self.active_token == Some(self.generation)
                    && self.desired_running
                    && self.failure.is_none()
            }
            AppPhase::Online => {
                self.operation == LifecycleOperation::None
                    && self.active_token == Some(self.generation)
                    && self.desired_running
                    && self.online
                    && self.failure.is_none()
            }
            AppPhase::Offline => {
                self.operation == LifecycleOperation::None
                    && self.active_token == Some(self.generation)
                    && self.desired_running
                    && !self.online
                    && self.failure.is_none()
            }
            AppPhase::Stopping => {
                matches!(
                    self.operation,
                    LifecycleOperation::Stop | LifecycleOperation::Reconcile
                ) && self.active_token == Some(self.generation)
                    && !self.desired_running
                    && !self.online
                    && self.failure.is_none()
            }
            AppPhase::Failed => {
                self.operation == LifecycleOperation::None
                    && self.active_token.is_none()
                    && !self.desired_running
                    && !self.online
                    && self.failure.is_some()
            }
        };
        if coherent {
            Ok(())
        } else {
            Err("phase, operation, token, intent, connectivity, and failure disagree")
        }
    }
}

#[derive(Clone, Debug, Default)]
pub struct AppLifecycleMachine {
    snapshot: AppLifecycleSnapshot,
}

impl AppLifecycleMachine {
    /// Largest generation represented exactly by every Rust, JavaScript, and
    /// Dart refinement of this state machine.
    pub const MAX_SAFE_GENERATION: u64 = 9_007_199_254_740_991;

    pub fn new() -> Self {
        Self::default()
    }

    pub fn snapshot(&self) -> AppLifecycleSnapshot {
        self.snapshot
    }

    pub fn dispatch(&mut self, event: LifecycleEvent) -> TransitionOutcome {
        use AppPhase::*;
        use LifecycleEvent::*;

        let outcome = match event {
            StartRequested => {
                if self.snapshot.phase != Stopped {
                    TransitionOutcome::Rejected
                } else if self.snapshot.generation < Self::MAX_SAFE_GENERATION {
                    let token = self.snapshot.generation + 1;
                    self.snapshot = AppLifecycleSnapshot {
                        phase: Starting,
                        operation: LifecycleOperation::Start,
                        generation: token,
                        active_token: Some(token),
                        desired_running: true,
                        online: false,
                        failure: None,
                    };
                    TransitionOutcome::Applied
                } else {
                    TransitionOutcome::Rejected
                }
            }
            StartSucceeded { token } => match self.token_relation(token) {
                TokenRelation::Stale => TransitionOutcome::Stale,
                TokenRelation::Invalid => TransitionOutcome::Rejected,
                TokenRelation::Current if self.snapshot.phase != Starting => {
                    TransitionOutcome::Rejected
                }
                TokenRelation::Current => {
                    self.snapshot.phase = if self.snapshot.online {
                        Online
                    } else {
                        Offline
                    };
                    self.snapshot.operation = LifecycleOperation::None;
                    TransitionOutcome::Applied
                }
            },
            StartFailed { token } => match self.token_relation(token) {
                TokenRelation::Stale => TransitionOutcome::Stale,
                TokenRelation::Invalid => TransitionOutcome::Rejected,
                TokenRelation::Current if self.snapshot.phase != Starting => {
                    TransitionOutcome::Rejected
                }
                TokenRelation::Current => {
                    self.fail(LifecycleFailure::Start);
                    TransitionOutcome::Applied
                }
            },
            ConnectivityChanged { token, online } => match self.token_relation(token) {
                TokenRelation::Stale => TransitionOutcome::Stale,
                TokenRelation::Invalid => TransitionOutcome::Rejected,
                TokenRelation::Current
                    if !matches!(self.snapshot.phase, Starting | Online | Offline) =>
                {
                    TransitionOutcome::Rejected
                }
                TokenRelation::Current => {
                    let target = if self.snapshot.phase == Starting {
                        Starting
                    } else if online {
                        Online
                    } else {
                        Offline
                    };
                    if self.snapshot.online == online && self.snapshot.phase == target {
                        TransitionOutcome::Stuttered
                    } else {
                        self.snapshot.online = online;
                        self.snapshot.phase = target;
                        TransitionOutcome::Applied
                    }
                }
            },
            RuntimeFailed { token } => match self.token_relation(token) {
                TokenRelation::Stale => TransitionOutcome::Stale,
                TokenRelation::Invalid => TransitionOutcome::Rejected,
                TokenRelation::Current
                    if !matches!(self.snapshot.phase, Starting | Online | Offline) =>
                {
                    TransitionOutcome::Rejected
                }
                TokenRelation::Current => {
                    self.fail(LifecycleFailure::Runtime);
                    TransitionOutcome::Applied
                }
            },
            StopRequested => match self.snapshot.phase {
                Stopped | Stopping => TransitionOutcome::Stuttered,
                Failed => TransitionOutcome::Rejected,
                Starting | Online | Offline => self.begin_stop(LifecycleOperation::Stop),
            },
            StopSucceeded { token } => match self.token_relation(token) {
                TokenRelation::Stale => TransitionOutcome::Stale,
                TokenRelation::Invalid => TransitionOutcome::Rejected,
                TokenRelation::Current if self.snapshot.phase != Stopping => {
                    TransitionOutcome::Rejected
                }
                TokenRelation::Current => {
                    self.snapshot = AppLifecycleSnapshot {
                        generation: self.snapshot.generation,
                        ..AppLifecycleSnapshot::default()
                    };
                    TransitionOutcome::Applied
                }
            },
            StopFailed { token } => match self.token_relation(token) {
                TokenRelation::Stale => TransitionOutcome::Stale,
                TokenRelation::Invalid => TransitionOutcome::Rejected,
                TokenRelation::Current if self.snapshot.phase != Stopping => {
                    TransitionOutcome::Rejected
                }
                TokenRelation::Current => {
                    self.fail(LifecycleFailure::Stop);
                    TransitionOutcome::Applied
                }
            },
            ReconcileRequested => {
                if self.snapshot.phase == Failed {
                    self.begin_stop(LifecycleOperation::Reconcile)
                } else {
                    TransitionOutcome::Rejected
                }
            }
        };

        debug_assert!(self.snapshot.validate().is_ok());
        outcome
    }

    fn token_relation(&self, token: u64) -> TokenRelation {
        match (token, self.snapshot.active_token) {
            (0, _) => TokenRelation::Invalid,
            (candidate, Some(active)) if candidate == active => TokenRelation::Current,
            (candidate, _) if candidate <= self.snapshot.generation => TokenRelation::Stale,
            _ => TokenRelation::Invalid,
        }
    }

    fn begin_stop(&mut self, operation: LifecycleOperation) -> TransitionOutcome {
        if self.snapshot.generation >= Self::MAX_SAFE_GENERATION {
            return TransitionOutcome::Rejected;
        }
        let token = self.snapshot.generation + 1;
        self.snapshot = AppLifecycleSnapshot {
            phase: AppPhase::Stopping,
            operation,
            generation: token,
            active_token: Some(token),
            desired_running: false,
            online: false,
            failure: None,
        };
        TransitionOutcome::Applied
    }

    fn fail(&mut self, failure: LifecycleFailure) {
        self.snapshot = AppLifecycleSnapshot {
            phase: AppPhase::Failed,
            operation: LifecycleOperation::None,
            generation: self.snapshot.generation,
            active_token: None,
            desired_running: false,
            online: false,
            failure: Some(failure),
        };
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::collections::{HashSet, VecDeque};

    #[test]
    fn stale_start_completion_cannot_resurrect_a_stopping_session() {
        let mut machine = AppLifecycleMachine::new();
        assert_eq!(
            machine.dispatch(LifecycleEvent::StartRequested),
            TransitionOutcome::Applied
        );
        let start_token = machine.snapshot().active_token.unwrap();
        assert_eq!(
            machine.dispatch(LifecycleEvent::StopRequested),
            TransitionOutcome::Applied
        );
        assert_eq!(
            machine.dispatch(LifecycleEvent::StartSucceeded { token: start_token }),
            TransitionOutcome::Stale
        );
        assert_eq!(machine.snapshot().phase, AppPhase::Stopping);
    }

    #[test]
    fn only_previously_allocated_tokens_are_stale() {
        let mut machine = AppLifecycleMachine::new();
        assert_eq!(
            machine.dispatch(LifecycleEvent::StartRequested),
            TransitionOutcome::Applied
        );
        let snapshot = machine.snapshot();

        assert_eq!(
            machine.dispatch(LifecycleEvent::StartSucceeded { token: 0 }),
            TransitionOutcome::Rejected
        );
        assert_eq!(machine.snapshot(), snapshot);
        assert_eq!(
            machine.dispatch(LifecycleEvent::StartSucceeded { token: 2 }),
            TransitionOutcome::Rejected
        );
        assert_eq!(machine.snapshot(), snapshot);

        assert_eq!(
            machine.dispatch(LifecycleEvent::StopRequested),
            TransitionOutcome::Applied
        );
        let stopping = machine.snapshot();
        assert_eq!(
            machine.dispatch(LifecycleEvent::StartSucceeded { token: 1 }),
            TransitionOutcome::Stale
        );
        assert_eq!(machine.snapshot(), stopping);
    }

    #[test]
    fn malformed_foreign_snapshot_has_no_capabilities() {
        let malformed = AppLifecycleSnapshot {
            phase: AppPhase::Online,
            ..AppLifecycleSnapshot::default()
        };

        assert!(malformed.validate().is_err());
        assert_eq!(malformed.capabilities(), AppCapabilities::CLOSED);
    }

    #[test]
    fn generation_is_bounded_by_the_cross_runtime_exact_integer_range() {
        let at_limit = AppLifecycleSnapshot {
            generation: AppLifecycleMachine::MAX_SAFE_GENERATION,
            ..AppLifecycleSnapshot::default()
        };
        assert!(at_limit.validate().is_ok());
        let mut machine = AppLifecycleMachine { snapshot: at_limit };
        assert_eq!(
            machine.dispatch(LifecycleEvent::StartRequested),
            TransitionOutcome::Rejected
        );

        let beyond_limit = AppLifecycleSnapshot {
            generation: AppLifecycleMachine::MAX_SAFE_GENERATION + 1,
            ..AppLifecycleSnapshot::default()
        };
        assert!(beyond_limit.validate().is_err());
        assert_eq!(beyond_limit.capabilities(), AppCapabilities::CLOSED);
    }

    #[test]
    fn failure_requires_explicit_reconciliation_before_restart() {
        let mut machine = AppLifecycleMachine::new();
        machine.dispatch(LifecycleEvent::StartRequested);
        let token = machine.snapshot().active_token.unwrap();
        machine.dispatch(LifecycleEvent::StartFailed { token });
        assert_eq!(machine.snapshot().phase, AppPhase::Failed);
        assert_eq!(
            machine.dispatch(LifecycleEvent::StartRequested),
            TransitionOutcome::Rejected
        );
        assert_eq!(
            machine.dispatch(LifecycleEvent::ReconcileRequested),
            TransitionOutcome::Applied
        );
        let reconcile_token = machine.snapshot().active_token.unwrap();
        machine.dispatch(LifecycleEvent::StopSucceeded {
            token: reconcile_token,
        });
        assert_eq!(machine.snapshot().phase, AppPhase::Stopped);
    }

    #[test]
    fn bounded_event_graph_is_total_deterministic_and_invariant_safe() {
        let mut queue = VecDeque::from([AppLifecycleMachine::new()]);
        let mut seen = HashSet::new();
        let mut seen_phases = HashSet::new();

        while let Some(machine) = queue.pop_front() {
            let snapshot = machine.snapshot();
            let key = format!("{snapshot:?}");
            if !seen.insert(key) {
                continue;
            }
            seen_phases.insert(format!("{:?}", snapshot.phase));
            snapshot
                .validate()
                .expect("all reachable snapshots are coherent");
            if snapshot.generation >= 4 {
                continue;
            }
            let current = snapshot.active_token.unwrap_or(snapshot.generation);
            let stale = current.saturating_sub(1);
            let events = [
                LifecycleEvent::StartRequested,
                LifecycleEvent::StartSucceeded { token: current },
                LifecycleEvent::StartSucceeded { token: stale },
                LifecycleEvent::StartFailed { token: current },
                LifecycleEvent::ConnectivityChanged {
                    token: current,
                    online: true,
                },
                LifecycleEvent::ConnectivityChanged {
                    token: current,
                    online: false,
                },
                LifecycleEvent::RuntimeFailed { token: current },
                LifecycleEvent::StopRequested,
                LifecycleEvent::StopSucceeded { token: current },
                LifecycleEvent::StopFailed { token: current },
                LifecycleEvent::ReconcileRequested,
            ];
            for event in events {
                let mut next = machine.clone();
                let first = next.dispatch(event);
                let next_snapshot = next.snapshot();
                let mut replay = machine.clone();
                let second = replay.dispatch(event);
                assert_eq!(first, second, "outcome must be deterministic for {event:?}");
                assert_eq!(
                    next_snapshot,
                    replay.snapshot(),
                    "next state must be deterministic for {event:?}"
                );
                queue.push_back(next);
            }
        }

        for phase in [
            "Stopped", "Starting", "Online", "Offline", "Stopping", "Failed",
        ] {
            assert!(
                seen_phases.contains(phase),
                "bounded graph must reach {phase}"
            );
        }
    }
}
